src/System.Security.Claims/src/System/Security/Claims/Claim.cs

```diff
@@ -0,0 +1,550 @@
+
+            int numberOfPropertiesWritten = 1;
+            SerializationMask mask = SerializationMask.None;
+            if (string.Equals(_type, ClaimsIdentity.DefaultNameClaimType))
+            {
+                mask |= SerializationMask.NameClaimType;
+            }
+            else if (string.Equals(_type, ClaimsIdentity.DefaultRoleClaimType))
+            {
+                mask |= SerializationMask.RoleClaimType;
+            }
+            else
+            {
+                numberOfPropertiesWritten++;
+            }
+
+            if (!string.Equals(_valueType, ClaimValueTypes.String, StringComparison.Ordinal))
+            {
+                numberOfPropertiesWritten++;
+                mask |= SerializationMask.StringType;
+            }
+
+            if (!string.Equals(_issuer, ClaimsIdentity.DefaultIssuer, StringComparison.Ordinal))
+            {
+                numberOfPropertiesWritten++;
+                mask |= SerializationMask.Issuer;
+            }
+
+            if (string.Equals(_originalIssuer, _issuer, StringComparison.Ordinal))
+            {
+                mask |= SerializationMask.OriginalIssuerEqualsIssuer;
+            }
+            else if (!string.Equals(_originalIssuer, ClaimsIdentity.DefaultIssuer, StringComparison.Ordinal))
+            {
+                numberOfPropertiesWritten++;
+                mask |= SerializationMask.OriginalIssuer;
+            }
+
+            if (Properties.Count > 0)
+            {
+                numberOfPropertiesWritten++;
+                mask |= SerializationMask.HasProperties;
+            }
+
+            // TODO - brentschmaltz - maximum size ??
+            if (userData != null && userData.Length > 0)
+            {
+                numberOfPropertiesWritten++;
+                mask |= SerializationMask.UserData;
+            }
+
+            writer.Write((Int32)mask);
+            writer.Write((Int32)numberOfPropertiesWritten);
+            writer.Write(_value);
+
+            if (((mask & SerializationMask.NameClaimType) != SerializationMask.NameClaimType) && ((mask & SerializationMask.RoleClaimType) != SerializationMask.RoleClaimType))
+            {
+                writer.Write(_type);
+            }
+
+            if ((mask & SerializationMask.StringType) == SerializationMask.StringType)
+            {
+                writer.Write(_valueType);
+            }
+
+            if ((mask & SerializationMask.Issuer) == SerializationMask.Issuer)
+            {
+                writer.Write(_issuer);
+            }
+
+            if ((mask & SerializationMask.OriginalIssuer) == SerializationMask.OriginalIssuer)
+            {
+                writer.Write(_originalIssuer);
+            }
+
+            if ((mask & SerializationMask.HasProperties) == SerializationMask.HasProperties)
+            {
+                writer.Write(Properties.Count);
+                foreach (var key in Properties.Keys)
```

Comment:
Again, this will allocate a `KeysCollection<TKey>` instance. This and the enumerator allocation can be avoided by `foreach`ing over `_properties` directly, to use Dictionary's struct enumerator, e.g.:

```c#
foreach (var pair in _properties)
{
    writer.Write(pair.Key);
    writer.Write(pair.Value);
}
```